### ISSUE TITLE
Question: Partition Level Consumers

### ISSUE THREAD

**Issue by annymsMthd** [confluentinc/confluent-kafka-dotnet#790]
Description
===========
Are there any plans to provide an API for parition level consumers for a consumer group? Much like this library for go 
https://github.com/bsm/sarama-cluster/blob/master/README.md

**Comment by mhowlett** [comment]
You don't need to use the consumer group functionality, you can call `Consumer.Assign` directly on the partitions you want to read from.

**Comment by annymsMthd** [comment]
Do you then lose the ability to have the partition balancing and offset tracking a consumer group gives you?

**Comment by mhowlett** [comment]
you can use the commit offsets api, even if you're not using the consumer group balancing feature.

if you are `Assign`ed to more than one partition, you can't control which partition from the assigned set the next message should be consumed from (at a quick glance, it appears sarama allows this, and is what you mean by partition-level consumers).

what's your use case?

**Comment by mhowlett** [comment]
Actually, different API, but you could use `Pause` and `Resume` to achieve this. Forgot about those. Still interested in your use case.

**Comment by annymsMthd** [comment]
The use case is working through a topic keyed by something like VehicleId. I have the guarantee that those messages will be in order by partition. I would like to have a group that I could assign that would work through the messages in each partition in order. This could be achieved by deploying enough processes to match the partition count but it would be nice if a consumer was assigned two partitions I count work through each independently

**Comment by annymsMthd** [comment]
This would also be a good basis for building more advanced streaming capabilities.

**Comment by wdamron** [comment]
The ability to independently consume from assigned partitions (using consumer groups) can be helpful when a consumer is doing some latency-bound work for each message and/or uses retry logic (common for at-least-once processing).

With the current shared queue, a slow/blocking consumer causes head-of-line blocking for all its assigned partitions. If the consumer blocks for long enough, of course the max poll interval will be exceeded and partitions will be rebalanced.

It seems partition-level consumers are supported by librdkafka [(reference)](https://github.com/edenhill/librdkafka/wiki/FAQ#what-are-partition-queues-and-why-are-some-partitions-slower-than-others), but exposing a clean API around them is probably easier said than done.

In C, the setup might be something roughly like the following:

```
// get partition queue:
rkqu = rd_kafka_queue_get_partition(rk, topic, partition);
// disable forwarding to main queue:
rd_kafka_queue_forward(rkqu, NULL);

// consume partition:
rkm = rd_kafka_consume_queue(rkqu, timeout);

// elsewhere, call poll to ensure callbacks are handled:
rkm = rd_kafka_consumer_poll(rk, timeout); // will always return null
```

Maybe during the partitions-assigned callback, there could be functionality to get the partition queue while disabling its forwarding, then call `.Consume` in a loop (thread, Task, etc).

**Comment by wdamron** [comment]
A basic use-case could be a consumer which calls out to an external service or database for each message, retrying for some amount of time (say 60 seconds) before forwarding the message to a dead-letter queue and continuing on. While a message is being retried, the consumer would be blocked for all its assigned partitions, despite only 1 partition being blocked by that message. 

Partition-level consumers would work around this somewhat, allowing the "slow" partition to not block the others, and also allowing the main `.Consume` method to be called more often than the occasionally-blocked handler can call it (ensuring callbacks are handled).

**Comment by jmathai-confluent** [comment]
If you have questions about partition assignment with current versions, please open a new issue.
